### PR TITLE
fix(platform-server): fix an exception when HostListener('window:scroll') is used on the server

### LINKED ISSUE #15000: bug(platform-server): @HostListener('window') doesn't work
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
```

**Current behavior**
Adding
```ts
  @HostListener('window:scroll')
  track() {
    console.log('scroll');
  }
```

will give this error
```sh
EXCEPTION: Uncaught (in promise): TypeError: Cannot read property '_eventListenersMap' of undefined
TypeError: Cannot read property '_eventListenersMap' of undefined
    at Parse5DomAdapter.on (/Users/jjean/projects/ng-universal-demo/dist/server.js:35214:50)
    at Parse5DomAdapter.onAndCancel (/Users/jjean/projects/ng-universal-demo/dist/server.js:35232:18)
    at /Users/jjean/projects/ng-universal-demo/dist/server.js:36952:124
    at ZoneDelegate.invoke (/Users/jjean/projects/ng-universal-demo/dist/server.js:52558:26)
    at Zone.run (/Users/jjean/projects/ng-universal-demo/dist/server.js:52354:43)
    at NgZone.runOutsideAngular (/Users/jjean/projects/ng-universal-demo/dist/server.js:5339:31)
    at DefaultServerRendererV2.listen (/Users/jjean/projects/ng-universal-demo/dist/server.js:36951:32)
    at listenToElementOutputs (/Users/jjean/projects/ng-universal-demo/dist/server.js:11517:64)
    at createViewNodes (/Users/jjean/projects/ng-universal-demo/dist/server.js:14122:17)
    at createRootView (/Users/jjean/projects/ng-universal-demo/dist/server.js:14057:5)
    at Object.createProdRootView [as createRootView] (/Users/jjean/projects/ng-universal-demo/dist/server.js:14613:12)
    at ComponentFactory_.create (/Users/jjean/projects/ng-universal-demo/dist/server.js:11906:49)
    at ViewContainerRef_.createComponent (/Users/jjean/projects/ng-universal-demo/dist/server.js:12100:65)
    at RouterOutlet.activate (/Users/jjean/projects/ng-universal-demo/dist/server.js:60073:44)
    at ActivateRoutes.placeComponentIntoOutlet (/Users/jjean/projects/ng-universal-demo/dist/server.js:59266:20)
ORIGINAL STACKTRACE:
Error: Uncaught (in promise): TypeError: Cannot read property '_eventListenersMap' of undefined
TypeError: Cannot read property '_eventListenersMap' of undefined
    at Parse5DomAdapter.on (/Users/jjean/projects/ng-universal-demo/dist/server.js:35214:50)
    at Parse5DomAdapter.onAndCancel (/Users/jjean/projects/ng-universal-demo/dist/server.js:35232:18)
    at /Users/jjean/projects/ng-universal-demo/dist/server.js:36952:124
    at ZoneDelegate.invoke (/Users/jjean/projects/ng-universal-demo/dist/server.js:52558:26)
    at Zone.run (/Users/jjean/projects/ng-universal-demo/dist/server.js:52354:43)
    at NgZone.runOutsideAngular (/Users/jjean/projects/ng-universal-demo/dist/server.js:5339:31)
    at DefaultServerRendererV2.listen (/Users/jjean/projects/ng-universal-demo/dist/server.js:36951:32)
    at listenToElementOutputs (/Users/jjean/projects/ng-universal-demo/dist/server.js:11517:64)
    at createViewNodes (/Users/jjean/projects/ng-universal-demo/dist/server.js:14122:17)
    at createRootView (/Users/jjean/projects/ng-universal-demo/dist/server.js:14057:5)
    at Object.createProdRootView [as createRootView] (/Users/jjean/projects/ng-universal-demo/dist/server.js:14613:12)
    at ComponentFactory_.create (/Users/jjean/projects/ng-universal-demo/dist/server.js:11906:49)
    at ViewContainerRef_.createComponent (/Users/jjean/projects/ng-universal-demo/dist/server.js:12100:65)
    at RouterOutlet.activate (/Users/jjean/projects/ng-universal-demo/dist/server.js:60073:44)
    at ActivateRoutes.placeComponentIntoOutlet (/Users/jjean/projects/ng-universal-demo/dist/server.js:59266:20)
    at new Error (native)
    at resolvePromise (/Users/jjean/projects/ng-universal-demo/dist/server.js:52871:31) [angular]
    at resolvePromise (/Users/jjean/projects/ng-universal-demo/dist/server.js:52847:17) [angular]
    at /Users/jjean/projects/ng-universal-demo/dist/server.js:52907:17 [angular]
    at Object.onInvokeTask (/Users/jjean/projects/ng-universal-demo/dist/server.js:5388:41) [angular]
    at ZoneDelegate.invokeTask (/Users/jjean/projects/ng-universal-demo/dist/server.js:52590:36) [angular]
    at Zone.runTask (/Users/jjean/projects/ng-universal-demo/dist/server.js:52394:47) [<root> => angular]
    at drainMicroTaskQueue (/Users/jjean/projects/ng-universal-demo/dist/server.js:52757:35) [<root>]
    at ZoneTask.invoke (/Users/jjean/projects/ng-universal-demo/dist/server.js:52648:25) [<root>]
    at data.args.(anonymous function) (/Users/jjean/projects/ng-universal-demo/dist/server.js:53676:25) [<root>]
    at FSReqWrap.oncomplete (fs.js:123:15) [<root>]
```

**Expected behavior**
It should fail gracefully by doing nothing

**Minimal reproduction of the problem with instructions**
https://github.com/FrozenPandaz/ng-universal-demo/tree/host
```sh
npm install && npm start
```

**What is the motivation / use case for changing the behavior?**
Server should gracefully fail for things that it does not support

**Please tell us about your environment:**
Node 6.9+

* **Angular version:** 2.0.X
4.0.0-rc.2

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
Node
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
Typescript

### PR DESCRIPTION
**Pull request by vikerman**
Fixes #15000 

**Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
Trying to bootstrap when a component has @HostListener("window:scroll") causes an exception on the server

**What is the new behavior?**
No exception

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by alxhub — APPROVED** [review]

**Review by FrozenPandaz — APPROVED** [review]

**Comment by FrozenPandaz** [comment]
Do you want me to fix title, it can now query the document.head

File: modules/@angular/platform-server/src/parse5_adapter.ts [hunk 1]
```diff
@@ -31,7 +31,19 @@ function _notImplemented(methodName: string) {
  * Parses a document string to a Document object.
  */
 export function parseDocument(html: string) {
-  return parse5.parse(html, {treeAdapter: parse5.treeAdapters.htmlparser2});
+  let doc = parse5.parse(html, {treeAdapter: parse5.treeAdapters.htmlparser2});
+  let docElement = doc.childNodes[0];
+  for (let i = 0; i < docElement.childNodes.length; i++) {
+    let node = docElement.childNodes[i];
+    if (node.name === 'head') {
+      doc['head'] = node;
```
Comment (gdi2290) [inline modules/@angular/platform-server/src/parse5_adapter.ts]: this needs to be a getter for the head node

File: modules/@angular/platform-server/src/parse5_adapter.ts [hunk 2]
```diff
@@ -31,7 +31,19 @@ function _notImplemented(methodName: string) {
...
+    }
+    if (node.name === 'body') {
+      doc['body'] = node;
```
Comment (gdi2290) [inline modules/@angular/platform-server/src/parse5_adapter.ts]: this needs to be a getter for the body node

**Comment by gdi2290** [comment]
you're going to run into the same problem as universal 2.1.x had where there will be a case when you're referencing a detached node if the document was changed after creation. At the same time you should include `.title` where the bug I mentioned because more relevant

File: modules/@angular/platform-server/src/parse5_adapter.ts [hunk 3]
```diff
@@ -31,7 +31,19 @@ function _notImplemented(methodName: string) {
...
 export function parseDocument(html: string) {
-  return parse5.parse(html, {treeAdapter: parse5.treeAdapters.htmlparser2});
+  let doc = parse5.parse(html, {treeAdapter: parse5.treeAdapters.htmlparser2});
+  let docElement = doc.childNodes[0];
+  for (let i = 0; i < docElement.childNodes.length; i++) {
+    let node = docElement.childNodes[i];
+    if (node.name === 'head') {
+      doc['head'] = node;
+    }
+    if (node.name === 'body') {
```
Comment (alfaproject) [inline modules/@angular/platform-server/src/parse5_adapter.ts]: else if / switch

**Comment by jimjim2a** [comment]
There is another issue when the html string has a Doctype, it appears as the first node of the parsed document. Therefore the head and body nodes aren't retrieved properly.

**Comment by gdi2290** [comment]
@jimjim2a is correct this is also why we had weird searching logic for the rootNode
https://github.com/angular/universal/blob/master/modules/platform-node/node-document.ts

File: packages/platform-server/src/parse5_adapter.ts [hunk 4]
```diff
@@ -31,7 +31,19 @@ function _notImplemented(methodName: string) {
  * Parses a document string to a Document object.
  */
 export function parseDocument(html: string) {
-  return parse5.parse(html, {treeAdapter: parse5.treeAdapters.htmlparser2});
+  let doc = parse5.parse(html, {treeAdapter: parse5.treeAdapters.htmlparser2});
+  let docElement = doc.childNodes[0];
```
Comment (alxhub) [inline packages/platform-server/src/parse5_adapter.ts]: How does this interact with a `<!doctype>`?

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/platform-server/src/parse5_adapter.ts (hunk 4 above)
Comment (vikerman) [inline packages/platform-server/src/parse5_adapter.ts]: Actually all of them are children of the childNodes[0]. The top level document has a single childNode that is the document Element that has all other elements as children.

**Review by alxhub — APPROVED** [review]

**Review by alxhub — APPROVED** [review]

**Comment by chuckjaz** [comment]
This PR now needs to be rebased.

**Comment by vikerman** [comment]
@chuckjaz - Rebased

**Pull request merged**